### PR TITLE
Yurii/5263 matching engine request queue

### PR DESCRIPTION
**Pull request by durkmurder**
New logic for consuming/dispatching events from networking layer. A few key points about implementation:
- `matching.Engine` becomes `matching.Core` which is almost the same matching engine that was before but doesn't know anything about networking. 
- Introducing new `matching.Engine` which implements `Process` and other functions to receive data from networking layer. Purpose of `matching.Engine` is to collect events and dispatch them to `matching.Core` whenever there is a possibility. 
- `matching.Engine` and `matching.Core` run separate goroutines, this was done to prevent blocking of caller code and implement a nice processing pipeline. 
- Simplified flow: `Process(GoroutineA)` -> `process/filter/drop event by engine(GoroutineB)` -> `handle event by core(GoroutineC)`

### Beginning of pipeline
Processing starts by calling `Process` from networking layer. Internal channel is used to pass `*Event` to goroutine that serves events.
```golang
func (c *Engine) Process(originID flow.Identifier, event interface{}) error {
	c.pendingEventSink <- &Event{
		OriginID: originID,
		Msg:      event,
	}
	return nil
}
```
### Engine event handling
This goroutine is heart of event processing, it receives messages from internal events queue and when there are items that were pre-processed or filtered, dispatches them into corresponding sinks. `fetchEvent` is the simplest implementation but obviously we can implement any selecting logic that we want.
```golang
func (c *Engine) processEvents() {
	fetchEvent := func(queue *deque.Deque, sink EventSink) (*Event, EventSink) {
		event, ok := queue.PopFront()
		if !ok {
			return nil, nil
		}
		return event.(*Event), sink
	}

	for {
		pendingReceipt, receiptSink := fetchEvent(&c.pendingReceipts, c.receiptSink)
		pendingApproval, approvalSink := fetchEvent(&c.pendingApprovals, c.approvalSink)
		pendingApprovalResponse, approvalResponseSink := fetchEvent(&c.pendingApprovalResponses, c.approvalResponseSink)
		select {
		case event := <-c.pendingEventSink:
			c.processPendingEvent(event)
		case receiptSink <- pendingReceipt:
			continue
		case approvalSink <- pendingApproval:
			continue
		case approvalResponseSink <- pendingApprovalResponse:
			continue
		case <-c.unit.Quit():
			return
		}
	}
}
```

`select` statement is the core condition why this processing goroutine doesn't block on recv or send. Proof that it works: 
https://play.golang.org/p/lWrDsSd7eY4. This is a small example that I have used to verify the model. 

### Processing event inside of engine
This function is called as handler when we received some data from event sink. It's important to be fast here, something that is not trivial shouldn't be here since it steals CPU cycles from event delivery. If we need more complex processing/filtering we can always add a new goroutine down the pipeline.
```golang
func (c *Engine) processPendingEvent(event *Event) {
	switch event.Msg.(type) {
	case *flow.ExecutionReceipt:
		c.engineMetrics.MessageReceived(metrics.EngineMatching, metrics.MessageExecutionReceipt)
		if c.pendingReceipts.Len() < defaultReceiptQueueCapacity {
			c.pendingReceipts.PushBack(event)
		}
	case *flow.ResultApproval:
		c.engineMetrics.MessageReceived(metrics.EngineMatching, metrics.MessageResultApproval)
		if c.requiredApprovalsForSealConstruction < 1 {
			// if we don't require approvals to construct a seal, don't even process approvals.
			return
		}
		if c.pendingApprovals.Len() < defaultApprovalQueueCapacity {
			c.pendingApprovals.PushBack(event)
		}
	case *messages.ApprovalResponse:
		c.engineMetrics.MessageReceived(metrics.EngineMatching, metrics.MessageResultApproval)
		if c.requiredApprovalsForSealConstruction < 1 {
			// if we don't require approvals to construct a seal, don't even process approvals.
			return
		}
		if c.pendingApprovalResponses.Len() < defaultApprovalResponseQueueCapacity {
			c.pendingApprovalResponses.PushBack(event)
		}
	}
}
```

### Consuming events by `Core`
Last but not least. Core will consume events from engine with this statement:
```golang
	for {
		select {
		case event := <-e.receiptEventProvider:
			processItem(event)
		case event := <-e.approvalEventProvider:
			processItem(event)
		case event := <-e.approvalResponseEventProvider:
			processItem(event)
		case <-checkSealingTicker:
			e.checkSealing()
		case <-e.unit.Quit():
			return
		}
	}
```
This happens in separate goroutine so we can do whatever we like here. Currently this code is synchronous but it's very flexible in terms of adding more workers and changing this consumer to be concurrent.

### CODE REVIEW

**Comment by AlexHentschel** [comment]
I like the way, how you are approaching the problem. Some thoughts:
* The matching engine now doesn't implement the engine interface anymore. So maybe we could reflect this in the names: 
   - rename `context` to `engine`
   - and maybe call what is now `engine` the `core` (or something similar 🤷‍♂️ )
* To me, it would be intuitive to combine all the concurrent elements in what is now called `context`: 
   * the thread serving the `core` could essentially be part of the `context` 
   * the `unit` could also move over to `context` 
   * the  `core` itself would then just be the _non_-concurrent business logic

File: engine/consensus/matching/engine.go
Comment (arrivets) [inline engine/consensus/matching/engine.go:46]:
```suggestion
	core                               *Core
```

File: engine/consensus/matching/engine.go
Comment (arrivets) [inline engine/consensus/matching/engine.go:82]: would it be worth making these buffered channels?
Otherwise pushing items to them will block until the core is done processing the event.
Not sure if this is expected behaviour.

File: engine/consensus/matching/engine.go
Comment (arrivets) [inline engine/consensus/matching/engine.go:137]: why do you need the sink here?

File: engine/consensus/matching/engine.go
Comment (arrivets) [inline engine/consensus/matching/engine.go:152]: this will block unless receiptSink is a buffered channel

File: engine/consensus/matching/engine.go
Comment (arrivets) [inline engine/consensus/matching/engine.go:171]: maybe it would be good to log when the queues are full

**Review by arrivets — CHANGES_REQUESTED** [review]
Fantastic!! Just a comment about the channels between engine and core.. I am wondering if they should be buffered to make the processing of events asynchronous between engine and core

File: engine/consensus/matching/engine.go
Comment (durkmurder) [reply engine/consensus/matching/engine.go:152]: That's the idea that it won't, with this select it will block on whole statement and will unblock in 3 cases:
1. We are quitting
2. We are able to send data
3. We are able to receive data. 

So even if we can't send data to core we are still able to get data from even sink.

File: engine/consensus/matching/engine.go
Comment (durkmurder) [reply engine/consensus/matching/engine.go:137]: Here we select item and sink, if there is no item then we shouldn't add it as part of select statement. That's just an easy way to get (item, sink) or (nil and nil).

File: engine/consensus/matching/engine.go
Comment (durkmurder) [reply engine/consensus/matching/engine.go:82]: While we can add some buffering, this object is designed to work in non blocking manner, it really depends on processPendingEvent. If it doesn't do anything heavy then we won't block anything.

File: engine/consensus/matching/engine.go
Comment (arrivets) [reply engine/consensus/matching/engine.go:152]: I am not sure about that. 
I think it wont process the next event unil receiptSink <- pendingReceipt unblocks, whcih happens only if core consumes the receipt immediately. 
But I might be wrong..would be interesting to confirm with a test

Comment (durkmurder) [reply engine/consensus/matching/engine.go:152]: I have updated PR description, you can find there a link with working example why this works.

Comment (arrivets) [reply engine/consensus/matching/engine.go:152]: Yeah thanks a lot for the example. I wasn't aware that one could do non-blocking sends like that!
Very nice!

**Review by arrivets — APPROVED** [review]

**Review by AlexHentschel — COMMENTED** [review]
Great, this looks like we are almost there. One last round of polishing, and we should be good to merge. 👏 . 

I made small PR https://github.com/onflow/flow-go/pull/376 targeting your branch with suggested revisions: 
* updates of goDoc and comments;
* renamed receiver variables for `matching.Core` to `c` and for `matching.Engine` to `e`

In addition, I have a few design suggestions (not part of my PR revisions). I'll give an overview in the following and also provided some more detailed comments in the code. 

#### Separation of Concern between `matching.Core` and `matching.Engine`

From my perspective, a slightly different separation of concerns would be more intuitive:
1. `matching.Core` contains only the core algorithm. It is _not_ concurrency safe and is probably best run in a single thread. 
2. The entire threading and message queueing is in `matching.Engine`. Essentially `matching.Engine` manages the inputs, external concurrency and internal thread for feeding the `Core` logic with inputs. 

Specifically, I would suggest to 
* move the methods `Ready`, `Done`, `processPendingEvents`, `processPendingEvent` from `Core` into the `matching.Engine`: https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/core.go#L151-L230 Thereby, `Core` would expose the entry points for the individual message types (i.e. `onReceipt`, `onApproval`)
* Document in `matching.Core` that it is _not_ concurrency safe and should be executed by a single thread 

The design would then follow the same paradigm which already worked very well for the HotStuff (where all the input handling, concurrency and threading was handled by the `EventLoop`)

#### Go routines accessing the `matching.Engine`
* I think it is totally fine to utilize the networking Layer's worker thread do the minimal work of pushing to the queues. Especially, I would suggest to do all these operations https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/engine.go#L167-L191 the body of https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/engine.go#L125 For practical purposes, this code should be non-blocking.

**Comment by AlexHentschel** [comment]
### Comment on related work
this will probably conflict with some of Leo's revisions https://github.com/onflow/flow-go/pull/363. Would suggest to wait until @zhangchiqing's PR is merged as his revisions should only affect `matching.Core`'s business logic 🤞 , which is only moved by this PR (from Engine to Core) but not modified.

File: engine/consensus/matching/engine.go
Comment (AlexHentschel) [reply engine/consensus/matching/engine.go:170]: Even better would be to monitor the queue size through a Prometheus metric.

**Comment by durkmurder** [comment]
> Great, this looks like we are almost there. One last round of polishing, and we should be good to merge.  .
> 
> I made small PR #376 targeting your branch with suggested revisions:
> 
> * updates of goDoc and comments;
> * renamed receiver variables for `matching.Core` to `c` and for `matching.Engine` to `e`
> 
> In addition, I have a few design suggestions (not part of my PR revisions). I'll give an overview in the following and also provided some more detailed comments in the code.
> 
> #### Separation of Concern between `matching.Core` and `matching.Engine`
> From my perspective, a slightly different separation of concerns would be more intuitive:
> 
> 1. `matching.Core` contains only the core algorithm. It is _not_ concurrency safe and is probably best run in a single thread.
> 2. The entire threading and message queueing is in `matching.Engine`. Essentially `matching.Engine` manages the inputs, external concurrency and internal thread for feeding the `Core` logic with inputs.
> 
> Specifically, I would suggest to
> 
> * move the methods `Ready`, `Done`, `processPendingEvents`, `processPendingEvent` from `Core` into the `matching.Engine`: https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/core.go#L151-L230
>    Thereby, `Core` would expose the entry points for the individual message types (i.e. `onReceipt`, `onApproval`)
> * Document in `matching.Core` that it is _not_ concurrency safe and should be executed by a single thread
> 
> The design would then follow the same paradigm which already worked very well for the HotStuff (where all the input handling, concurrency and threading was handled by the `EventLoop`)
> 
> #### Go routines accessing the `matching.Engine`
> * I think it is totally fine to utilize the networking Layer's worker thread do the minimal work of pushing to the queues. Especially, I would suggest to do all these operations https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/engine.go#L167-L191
>    the body of https://github.com/onflow/flow-go/blob/9a2bc57326ddd2497e32137630cc17720992bbce/engine/consensus/matching/engine.go#L125
>    For practical purposes, this code should be non-blocking.

I agree with almost everything except keeping Core single threaded. While it's totally reasonable now, it may be a problem in future since I have a feeling that we will need make big changes to make approvals/receipt handling in concurrent way. For now let's do it, but let's just keep in mind that we might change it in future.

File: engine/consensus/matching/engine.go
Comment (durkmurder) [reply engine/consensus/matching/engine.go]: the problem is that I need to pass `approvalConduit` which is created in `Engine` constructor.

**Review by zhangchiqing — APPROVED** [review]
Great work solving the problem that the match engine is processing events in a single threaded manner, which is blocking pending events to be processed, and further blocking the networking layer from delivering messages to other engine. 👏 

Nice improvement.

While this PR keeps the single threaded behavior unchanged, with the increased load from more result approvals, we might have a higher chance to face the problem that the processing speed is behind the increasing incoming message volume, which I’m also a bit concerned. 

That being said, I agree with @durkmurder that better not commit to single threaded. I think in the future the matching core needs to be concurrently processing a certain number of approvals and a certain number of receipts, while doing checkSealing on the side. So the match engine is just to ensure the core is not overwhelmed by placing the queue in front of it. However, this is not a problem we have seen now, let’s maybe keep as is for now.

File: utils/concurrentqueue/concurrentqueue.go
Comment (durkmurder) [reply utils/concurrentqueue/concurrentqueue.go:6]: it's more than production ready, it's a very solid library. You can check github. It's way more performant than double linked list.

File: engine/consensus/matching/engine.go
Comment (AlexHentschel) [inline engine/consensus/matching/engine.go:147]: I would switch those two cases. If we have requested approvals, they are likely blocking progress. We should process them with priority.

File: engine/consensus/matching/engine.go
Comment (AlexHentschel) [inline engine/consensus/matching/engine.go:174]: Why don't we put this logic into the queue?

File: engine/consensus/matching/engine.go
Comment (AlexHentschel) [inline engine/consensus/matching/engine.go:219]: We handle most cases here through `consumeSingleEvent`(except for the sealing check). As we don't pass any context to `consumeSingleEvent` has to re-check the types and recover the information we already have here. 

To me, this adds unnecessary complexity (personal stylistic preference). Hence, I would like to suggest:
* in each case, we know precisely, that type of event we are expecting in the queue 
* lets do an unchecked cast and feed it directly into the corresponding method in code. 

Furthermore, I would like to suggest to make the following changes to `Core`:
* export methods we use in the engine. This provides some guidance what methods should be used externally. 
* In my opinion, core should handle all the expected errors internally. Any error that is escalated to the engine can be taken as a fatal problem. 
   - for now, you could just move this into `Core` https://github.com/onflow/flow-go/blob/823315497f38e8d508cd4acf361ad1691fe077c0/engine/consensus/matching/engine.go#L250-L257
   - in case `consumeEvents` receives an error from `Core`, we assume that the node's internal state is corrupted. In that case, we `log.Fatal().Err(err)` and are done.

**Review by AlexHentschel — APPROVED** [review]
Thanks.  I have a few stylistic requests: 

#### `Core`:
* export methods we use in the engine. This provides some guidance what methods should be used externally. 
* In my opinion, core should handle all the expected errors internally. Any error that is escalated to the engine can be taken as a fatal problem. 
   - for now, you could just move this into `Core` https://github.com/onflow/flow-go/blob/823315497f38e8d508cd4acf361ad1691fe077c0/engine/consensus/matching/engine.go#L250-L257
   - in case `consumeEvents` receives an error from `Core`, we assume that the node's internal state is corrupted. In that case, we `log.Fatal().Err(err)` and are done.

**Review by AlexHentschel — APPROVED** [review]
Great Work. This is ready to be merged. 

I have a few optional suggestions in this PR (targets this branch): https://github.com/onflow/flow-go/pull/395 (mostly extended the GoDoc of `FifoQueue`). I tried to prepare the PR in such a way that you can directly merge it into your branch without and additional code changes necessary. 

Either with or without my suggestions, please feel free to merge this PR.

**Pull request merged**
